### ISSUE TITLE
[vcpkg|binarycache] don't use triplet file hash.

### ISSUE THREAD

**Issue by Neumann-A** [microsoft/vcpkg#19984]
So I have a a custom triplet with per port customization and this is a problem for the binary cache since if I add a new customization everything will be invalidated, removed and rebuild although only one port was effectively changed. 

I thus suggest: Instead of using the file hash of the triplet to determine the hash for the binary cache it would be better to get the variables set by the triplet and use those to define the triplet hash for a port. Since the triplet is already read numerous times for compiler hash checks it shouldn't add a lot of additional overhead. 

This would greatly improve my workflow if I have to touch the triplet file.

**Comment by ras0219-msft** [comment]
This will also need to detect changes to environment variables and I'm not sure how to make

```cmake
set(ENV{PATH} "$ENV{PATH}:$ENV{SOME_BASE_DIR}/magic")
```

machine independent without hashing the source code.

**Comment by dg0yt** [comment]
> This will also need to detect changes to environment variables ...

I'm not sure if it is *always* desired to detect particular changes to environment variables. And even with hashing the source, you don't hash the contents of the `$ENV{SOME_BASE_DIR}` *input*.

How about a triplet revision variable (not replacing the hash, but as an input)? It might even be overridden per port...

**Comment by autoantwort** [comment]
> This will also need to detect changes to environment variables and I'm not sure how to make
> 
> ```cmake
> set(ENV{PATH} "$ENV{PATH}:$ENV{SOME_BASE_DIR}/magic")
> ```
> 
> machine independent without hashing the source code.

The question is if you always want this. Maybe one changes the env var `SOME_BASE_DIR` to switch between versions of something, then you want different hashes. Maybe we should add something like `set(VCPKG_IGNORE_FOR_HASH "$ENV{SOME_BASE_DIR}")` so that the value(s) of `VCPKG_IGNORE_FOR_HASH` are removed from the value of every variable so that they are not included in the hash.  
Or we don't include `ENV` variables in the hash. If you want to include a value of the env to be hashed you have to set a local variable to the value.

**Comment by Neumann-A** [comment]
@JackBoosY Why did you mark this as `requires:repro`?

**Comment by autoantwort** [comment]
@Neumann-A If you use a selfcompiled vcpkg-tool you can now use https://github.com/microsoft/vcpkg-tool/pull/596

**Comment by Osyotr** [comment]
Maybe teach VCPKG to also include `port-settings/triplet-name.port-name.cmake` and hash its content too?

**Comment by DownerCase** [comment]
Ran into this the other day, added a customization for a single port and was sad to see the world rebuild. Ended up copying the port to be an overlay port instead, which is a rather less ergonomic...

**Comment by redbaron** [comment]
@Neumann-A ,
> solve it with an if checking if a per port customization file exists

do you propose triplets to implicitly include per-port customisation files ?

**Comment by Neumann-A** [comment]
> @Neumann-A ,
> > solve it with an if checking if a per port customization file exists
> 
> do you propose triplets to implicitly include per-port customisation files ?

Yes and then explicitly add that file to the hash for that port.
